### PR TITLE
DB Refactoring/Chain RollBack Endpoint/Pg DB RollBack/Documentation 

### PR DESCRIPTION
**Pull request by dbkbali**
- Incorporates changes from rollback code sprint.
- Issue fixes for https://github.com/ShyftNetwork/go-empyrean/issues/95, https://github.com/ShyftNetwork/go-empyrean/issues/83, & https://github.com/ShyftNetwork/go-empyrean/issues/78, & https://github.com/ShyftNetwork/go-empyrean/issues/77

- Docker Start/Stop issue causing corrupted chaindata has been fixed

### CODE REVIEW

**Review by dutterbutter — APPROVED** [review]

File: cmd/geth/consolecmd_test.go [hunk 1]
```diff
@@ -87,7 +87,7 @@ func TestIPCAttachWelcome(t *testing.T) {
 		"--port", "0", "--maxpeers", "0", "--nodiscover", "--nat", "none",
 		"--etherbase", coinbase, "--shh", "--ipcpath", ipc)
 
-	time.Sleep(2 * time.Second) // Simple way to wait for the RPC endpoint to open
+	time.Sleep(30 * time.Second) // Simple way to wait for the RPC endpoint to open
```
Comment (tim-br) [inline cmd/geth/consolecmd_test.go:90]: 30 seconds is a long time to wait for tests.

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -1248,8 +1255,7 @@ func MakeChain(ctx *cli.Context, stack *node.Node) (chain *core.BlockChain, chai
 		cache.TrieNodeLimit = ctx.GlobalInt(CacheFlag.Name) * ctx.GlobalInt(CacheGCFlag.Name) / 100
 	}
 	vmcfg := vm.Config{EnablePreimageRecording: ctx.GlobalBool(VMEnableDebugFlag.Name)}
-	
-	fmt.Println("Calling NewBlock CHAIN in flags.go ******************************")
+
```
Comment (tim-br) [inline cmd/utils/flags.go]: remove new line

File: core/blockchain.go [hunk 3]
```diff
@@ -20,6 +20,7 @@ package core
 import (
 	"errors"
 	"fmt"
+
```
Comment (tim-br) [inline core/blockchain.go]: remove new line

File: core/blockchain_test.go [hunk 4]
```diff
@@ -24,8 +24,11 @@ import (
 	"testing"
 	"time"
 
+	"encoding/json"
+
```
Comment (tim-br) [inline core/blockchain_test.go]: new line

File: core/blockchain_test.go [hunk 5]
```diff
@@ -514,7 +545,10 @@ func TestBlocksInsertNonceError(t *testing.T)  { testInsertNonceError(t, true) }
 func testInsertNonceError(t *testing.T, full bool) {
 	for i := 1; i < 25 && !t.Failed(); i++ {
 		// Create a pristine chain and database
+		//@Shyft Note: Truncate Posgres Data Tables To Allow Reuse of Test Data
+		TruncateTables()
 		db, blockchain, err := newCanonical(ethash.NewFaker(), 0, full)
+
```
Comment (tim-br) [inline core/blockchain_test.go]: new line

File: core/state/statedb.go [hunk 6]
```diff
@@ -512,6 +512,8 @@ func (self *StateDB) RevertToSnapshot(revid int) {
 		self.journal[i].undo(self)
 	}
 	self.journal = self.journal[:snapshot]
+	//fmt.Printf("%+v\n", self.journal)
+	//fmt.Println(reflect.TypeOf(self.journal))
```
Comment (tim-br) [inline core/state/statedb.go]: rm unused code

File: eth/backend.go [hunk 7]
```diff
@@ -147,7 +150,7 @@ func New(ctx *node.ServiceContext, config *Config) (*Ethereum, error) {
 		return nil, genesisErr
 	}
 	log.Info("Initialised chain configuration", "config", chainConfig)
-	fmt.Println(config.GasPrice)
+	// fmt.Println(config.GasPrice)
```
Comment (tim-br) [inline eth/backend.go]: rm unused println

File: eth/sync_test.go [hunk 8]
```diff
@@ -17,15 +17,27 @@
 package eth
 
 import (
+	"os"
 	"sync/atomic"
 	"testing"
 	"time"
 
+	"github.com/ShyftNetwork/go-empyrean/shyfttest"
+
```
Comment (tim-br) [inline eth/sync_test.go]: new line

File: light/odr_test.go [hunk 9]
```diff
@@ -239,6 +271,7 @@ func testChainOdr(t *testing.T, protocol int, fn odrTestFn) {
 	)
 	gspec.MustCommit(ldb)
 	// Assemble the test environment
+
```
Comment (tim-br) [inline light/odr_test.go]: new line

File: shyft-cli/startShyftGeth.sh [hunk 10]
```diff
@@ -2,5 +2,5 @@
...
 else
-  /bin/geth --config config.toml --ws --wsaddr="0.0.0.0" --wsorigins "*" --nat=any --mine --minerthreads 4 --targetgaslimit 80000000 --unlock "0x43EC6d0942f7fAeF069F7F63D0384a27f529B062,0x9e602164C5826ebb5A6B68E4AFD9Cd466043dc4A,0x5Bd738164C61FB50eb12E227846CbaeF2dE965Aa,0xC04eE4131895F1d0C294D508AF65D94060AA42BB,0x07D899C4aC0c1725C35C5f816e60273B33a964F7" --password ./unlockPasswords.txt
+  /bin/geth --config config.toml --gcmode archive --ws --wsaddr="0.0.0.0" --wsorigins "*" --nat=any --mine --minerthreads 4 --targetgaslimit 80000000 --unlock "0x43EC6d0942f7fAeF069F7F63D0384a27f529B062,0x9e602164C5826ebb5A6B68E4AFD9Cd466043dc4A,0x5Bd738164C61FB50eb12E227846CbaeF2dE965Aa,0xC04eE4131895F1d0C294D508AF65D94060AA42BB,0x07D899C4aC0c1725C35C5f816e60273B33a964F7" --password ./unlockPasswords.txt
```
Comment (tim-br) [inline shyft-cli/startShyftGeth.sh:5]: what is gcmode archive and can we add it to config.toml? Can we also add the other cli flags to config.toml?

File: cmd/utils/flags.go (hunk 2 above)
Comment (dutterbutter) [reply cmd/utils/flags.go]: removed

File: core/blockchain_test.go (hunk 4 above)
Comment (dutterbutter) [reply core/blockchain_test.go]: removed

File: cmd/geth/consolecmd_test.go (hunk 1 above)
Comment (dutterbutter) [reply cmd/geth/consolecmd_test.go:90]: Interestingly, in order for the test to pass we needed it to be set at 30 seconds, which seems to vary machine to machine. At 2 seconds, the test fails due to a context time out.

**Review by GregTheGreek — APPROVED** [review]

**Pull request merged**